### PR TITLE
Remove index audit output type

### LINKED ISSUE #29881: Remove index audit trail in favor of filebeat
*Original comment by @albertzaharovits:*

Right now Security has the capability to index the audit events trail on the local or a remote cluster.
**I want to moot the removal of this in favor of filebeat.**
The standing logging audit trail will remain the only auditing option. It could possibly be improved with a syslog appender. 
**This way, logging audit trail will handle the durability of the audit events, and we will leave search-ability to the user to do.**

Given the overlapping of functionality with filebeat, the index auditing is more of a convenience functionality. More importantly, the asynchronicity of enqueuing audit events, shipping over the network and then indexing will make the [failure-state-when-auditing-unavailable](https://github.com/elastic/elasticsearch/issues/29905) feature impractical in this case. Another feature, namely [integrity-verification](https://github.com/elastic/elasticsearch/issues/29718) will also be tricky to implement, although, arguably, there's no clear solution for logging too (possibly ship events to syslog or Event Viewer and let them handle the integrity feature).
Another inconvenience is that we are indexing all the event's fields with fixed mappings, and the user cannot change this. The audit trail corpus is beefy, we should allow for indexing only a subset of fields with custom mappings.
Spice all this with the problem of permission around the audit index which is hard to get right in the remote cluster case.

Index audit trail has grown into a toy functionality which a security conscious user (presumably ones that enable audit logging in the first place) will want to stay away from.

TODO: In the process we ought to make audit events JSON docs.

CC @elastic/es-security 
pinging @clintongormley as this is a xstack venture

### PR DESCRIPTION
**Pull request by albertzaharovits**
Supersedes #37301

This PR removes the Index Audit Output type, following its deprecation in 6.7 by https://github.com/elastic/elasticsearch/pull/37671 . It also adds the migration notice (settings notice).

Closes #29881

### CODE REVIEW

File: docs/reference/migration/migrate_7_0/settings.asciidoc [hunk 1]
```diff
@@ -154,3 +154,19 @@ node's keystore, i.e., they are not to be specified via the cluster settings API
...
+==== Audit index output type removed
+
+All the settings under the `xpack.security.audit.index` namespace have been
+removed. In addition, the `xpack.security.audit.outputs` setting has been
+removed as well.
+
+These settings enabled and configured the audit index output type. This output
+type has been removed because it was unreliable in certain scenarios and this
+could have lead to dropping audit events while the operations on the system
+were allowed to continue as usual. This is a terrible failure state for an
```
Comment (jaymode) [inline docs/reference/migration/migrate_7_0/settings.asciidoc]: Can we leave out `This is a terrible failure state for an audit system which brought about its demise.` ?

File: docs/reference/settings/audit-settings.asciidoc [hunk 2]
```diff
@@ -14,38 +14,26 @@ file. For more information, see
...
-cluster or a separate cluster.
+This puts the auditing events in a dedicated file named `<clustername>_audit.log`
+on each node.
 +
-For backwards compatibility reasons, if you use the logfile output type, a
-`<clustername>_access.log` file is also created. It contains the same
-information, but it uses the older (pre-6.5.0) formatting style.
-If the backwards compatible format is not required, it should be disabled.
-To do that, change its logger level to `off` in the `log4j2.properties` file.
+For backwards compatibility reasons, a `<clustername>_access.log` file is also
```
Comment (jaymode) [inline docs/reference/settings/audit-settings.asciidoc]: These are master docs, so I think this doesn't apply anymore?

File: x-pack/docs/en/security/auditing/overview.asciidoc [hunk 3]
```diff
@@ -13,26 +13,6 @@ Audit logs are **disabled** by default. To enable this functionality, you
...
-----------------------------
-xpack.security.audit.outputs: [ index, logfile ]
-----------------------------
-
-TIP: If you choose to enable the `index` output type, we strongly recommend that 
-you still use the `logfile` output as the official record of events. If the 
-target index is unavailable (for example, during a rolling upgrade), the `index` 
-output can lose messages.
+The audit log persists events to a dedicated `<clustername>_audit.log` file on
+the host's file system (on each cluster node). For backwards compatibility
```
Comment (jaymode) [inline x-pack/docs/en/security/auditing/overview.asciidoc]:
```suggestion
the host's file system (on each node).
```

File: x-pack/docs/en/security/auditing/overview.asciidoc [hunk 4]
```diff
@@ -13,26 +13,6 @@ Audit logs are **disabled** by default. To enable this functionality, you
...
+reasons, a file named `<clustername>_access.log` is also generated.
```
Comment (jaymode) [inline x-pack/docs/en/security/auditing/overview.asciidoc]:
```suggestion
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 5]
```diff
@@ -402,31 +375,13 @@ protected Clock getClock() {
         securityContext.set(new SecurityContext(settings, threadPool.getThreadContext()));
         components.add(securityContext.get());
 
-        // audit trails construction
-        Set<AuditTrail> auditTrails = new LinkedHashSet<>();
+        // audit trail service construction
+        final List<AuditTrail> auditTrails = new ArrayList<>();
```
Comment (jaymode) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: no need to allocate a list for this. Just do something like:

```suggestion
        final List<AuditTrail> auditTrails = XPackSettings.AUDIT_ENABLED.get(settings) ? 
            Collections.singletonList(new LoggingAuditTrail(settings, clusterService, threadPool)) : 
            Collections.emptyList();
```

Note: this is probably poorly formatted.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 6]
```diff
@@ -998,23 +865,8 @@ public AsyncSender interceptSender(AsyncSender sender) {
     @Override
     public UnaryOperator<Map<String, IndexTemplateMetaData>> getIndexTemplateMetaDataUpgrader() {
         return templates -> {
+            // .security index is not managed by using templates anymore
             templates.remove(SECURITY_TEMPLATE_NAME);
```
Comment (jaymode) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]:
```suggestion
             templates.remove(SECURITY_TEMPLATE_NAME);
             templates.remove("security_audit_log");
```

**Review by jaymode — CHANGES_REQUESTED** [review]
nice stats. I left some comments

File: docs/reference/settings/audit-settings.asciidoc (hunk 2 above)
Comment (albertzaharovits) [reply docs/reference/settings/audit-settings.asciidoc]: You're right 🙂  At first, this made me believe I still had a PR to do to remove the bwc audit format, that's how much I trust our docs!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java (hunk 6 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: good catch!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 7]
```diff
@@ -376,9 +376,9 @@ protected Clock getClock() {
         components.add(securityContext.get());
 
         // audit trail service construction
-        final List<AuditTrail> auditTrails = XPackSettings.AUDIT_ENABLED.get(settings) ? 
-            Collections.singletonList(new LoggingAuditTrail(settings, clusterService, threadPool)) : 
-            Collections.emptyList();
+        final List<AuditTrail> auditTrails = XPackSettings.AUDIT_ENABLED.get(settings)
+                ? Collections.singletonList(new LoggingAuditTrail(settings, clusterService, threadPool))
+                : Collections.emptyList();
         if (XPackSettings.AUDIT_ENABLED.get(settings)) {
```
Comment (jaymode) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: we should get rid of this block

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 8]
```diff
@@ -998,23 +867,9 @@ public AsyncSender interceptSender(AsyncSender sender) {
...
-                        .settings(IndexAuditTrail.customAuditIndexSettings(settings, logger))
-                        .build();
-                templates.put(IndexAuditTrail.INDEX_TEMPLATE_NAME, auditMetadata);
-            } catch (IOException e) {
-                // TODO: should we handle this with a thrown exception?
-                logger.error("Error loading template [{}] as part of metadata upgrading", IndexAuditTrail.INDEX_TEMPLATE_NAME);
-            }
-
+            // .security index is not managed by using templates anymore
+             templates.remove(SECURITY_TEMPLATE_NAME);
```
Comment (jaymode) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: looks like an extra space got added before these lines in the PR review. Can you fix this?

**Review by jaymode — APPROVED** [review]
I left two minor comments. Otherwise LGTM

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java (hunk 8 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: eagle eyes!

**Comment by albertzaharovits** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by albertzaharovits** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by albertzaharovits** [comment]
> 18:58:42     FAILURE: Build failed with an exception.
18:58:42 
18:58:42     * Where:
18:58:42     Build file '/tmp/junit14158968466601980319/build.gradle' line: 48
18:58:42 
18:58:42     * What went wrong:
18:58:42     A problem occurred evaluating root project 'junit14158968466601980319'.
18:58:42     > Could not get unknown property 'unitTest' for root project 'junit14158968466601980319' of type org.gradle.api.Project.

@elasticmachine run elasticsearch-ci/1

**Comment by albertzaharovits** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

**Review by tvernum — APPROVED** [review]
LGTM.
We could probably do a followup to simplify/remove `AuditTrail` etc, but that's not needed here.

**Pull request merged**